### DISCUSSION TITLE
how does `localePath()` work with dynamically generated routes?

### DISCUSSION THREAD

**Discussion by katerlouis** [nuxt-modules/i18n#1347] [Q&A]
I'm building a multilang site using storyblok as cms with `prefix_except_default` strategy. Before using storyblok, I used `localePath('some-route-name')` (emphasis on route-**name** here!) – But now that we generate **most** pages with storyblok, I don't know how to get the localePath for such storyblok page, when I need to reference it hardcoded somehwere. Again: some pages are still completely custom coded, but require links to storyblok pages. 

Any suggestions on that?

**Comment by rchl** [comment]
I believe you want https://i18n.nuxtjs.org/lang-switcher#dynamic-route-parameters

**Reply by katerlouis** [reply]
> I believe you want https://i18n.nuxtjs.org/lang-switcher#dynamic-route-parameters

Thanks! But I don't think this is what I'm looking for. The dynamic routes have already been translated in our cms, storyblok. Maybe an explicit example makes it more clear.

Let's take "imprint" for instance. In our default language the url is `dumbo.design/impressum`.
![image](https://user-images.githubusercontent.com/1983382/143407239-4382b053-85bd-4232-a50c-8db414e9fec4.png)

Storyblok now allows to set another slug, leaving us with `dumbo.design/en/imprint`
![image](https://user-images.githubusercontent.com/1983382/143407281-fabe6298-c074-4fe2-a3b1-809c661cab99.png)

In Nuxt I extended the router to catch all routes and check if there's a storyblok story for that.
```js
// nuxt.config.js
router: {
  middleware: ['dursor-reset'],

  extendRoutes(routes, resolve) {
    routes.push({
      name: 'storyblok-funnel',
      path: '/*',
      component: resolve(__dirname, 'src/components/storyblok/story.vue'),
    });
  },
},
```

In this routes component I try to fetch the story and redirect to 404 if it's not present.
```js
// src/components/storyblok/story.vue
async asyncData({ $storyblok, route, redirect, localePath }) {
  try {
    const story = await $storyblok.getStory(route.params.pathMatch); // pathMatch does not contain locale

    return {
      story,
    };
  }

  catch (error) {
    console.error(error);
    redirect(localePath({ name: '404' }));
  }
},
```

`$storyblok.getStory()` also passes the current locale
```js
async getStory(path, options = {}) {
  const result = await $storyapi.get(`cdn/stories/${path}`, {
    language: i18n.locale,
    resolve_links: 'url',
    resolve_relations: resolveRelations, // globally resolve these relations
    version: this.version(),
    ...options,
  });

  return result.data.story;
},
```

Up to this point everything works flawlessly and I'm really happy. The issue now is when I want to hardcode an imprint link in custom components like the `<Footer>` for instance. `localePath('imprint')` does not exist anymore. In fact, there are no storyblok routes what so ever in `.nuxt/router.js`, which makes sense of course. But I guess that's what we need for `localePath()` to work, right? 

Is there a way to write all story-routes into `.nuxt/router.js` on server-start?
Or do you have any other suggestion on how to manage translated urls from storyblok?

**Reply by rchl** [reply]
Hardcoding a link to a storyblock route should work but would require different syntax. I believe in this specific case it would be `localePath({ name: 'storyblok-funnel', params: { 0: 'imprint' })`. Or maybe `localePath('/imprint')` would also work but I always discourage using the path variant.

> Is there a way to write all story-routes into .nuxt/router.js on server-start?

If Storyblock provides an API to get all routes then I guess you could just do an API call and generate all routes in `nuxt.config.js` (remember that nuxt config can return an async function).

**Reply by katerlouis** [reply]
Thanks for the really quick reply!
I've tried that before, and obviously in default locale `localePath('/impressum')` works just fine. But see what happens when the locale is `en`

![image](https://user-images.githubusercontent.com/1983382/143411931-b1c0da13-5811-4c7c-be2c-04b1f222c3d3.png)

I get `/en/impressum` but expect `/en/imprint`

Exporting the entire config in an async function seems a bit overkill and unfortunately I don't have access to the nuxt context to use `$storyblok` plugin inside that async function.

My usecase cannot be this uncommon, right? What am I missing? I don't have to use `localePath` necessarily. If there's another solution I'm happy to oblige :D – What's important to me is: only using one function when hardcoding `<NuxtLink :to="someFunc('impressum')">`

**Reply by katerlouis** [reply]
Maybe I can use some form of hook in nuxt-i18n to extend the routes with access to nuxts `context`? nuxt-i18n obviously also extends the routes outside of `nuxt.config.js`

**Reply by katerlouis** [reply]
Oh, I just noticed that when doing 
```js
extendRoutes(routes, resolve) {
  routes.push({
    name: 'pilz',
    path: '/pilz',
    component...
}), 
```

that nuxt-i18n hooks into that and already creates `pilz___de` and `pilz___en` routes in `.nuxt/router.js` – but how can I provide the translated path? `pilz___en` gets `/en/pilz`, but would need the translated `/en/mushroom`

**Reply by rchl** [reply]
> I get /en/impressum but expect /en/imprint

Yes, I see now. So only Storyblock knows the mapping route -> localized name and so to make `localePath` work correctly, the only solution is to make nuxt-i18n aware of the mapping. Currently the only way to do that dynamically is using by using https://i18n.nuxtjs.org/lang-switcher#dynamic-route-parameters but that supports the currently opened route only. Maybe we are missing something like that but for multiple routes at once. Then you could create a plugin that sets it once for all routes.

> Maybe I can use some form of hook in nuxt-i18n to extend the routes with access to nuxts context? nuxt-i18n obviously also extends the routes outside of nuxt.config.js

Nuxt extends the routes but there is no access to `context` at that point.

**Reply by rchl** [reply]
> that nuxt-i18n hooks into that and already creates `pilz___de` and `pilz___en` routes in `.nuxt/router.js` – but how can I provide the translated path? `pilz___en` gets `/en/pilz`, but would need the translated `/en/mushroom`

https://i18n.nuxtjs.org/options-reference#pages is for providing the mapping but you don't have access to `context` at that point.

**Reply by katerlouis** [reply]
So with `parsePages: true` nuxt-i18n will look through files in `pages.vue`, serach for `nuxtI18n.paths` and push the correct routes accordingly? 

`pages: {}` is only used if `parsePages` is `false`? So no matter what, this doesn't help me, since I intend to use `nuxtI18n.paths` for some pages that don't come from storyblok :D

It's hard to wrap my head around all this; please tell me if my assumption is correct, that `extendRoutes()` is picked up by nuxt-i18n and takes each route to create new routes for every locale specifically? 

Could you maybe provide an example using `pages: {}` to realize my `impressum` use case? I'm thinking that fetching all routes without the context isn't too bad. All I need is axios and my token.

**Reply by rchl** [reply]
Something like this I guess:
```
  pages: {
    impressum: {
      en: '/imprint',
    }
  }
```
